### PR TITLE
wasm-component-ld 0.5.10 (new formula), wasi-runtimes: depend on `wasm-component-ld`

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This is needed to enable the `rust` formula to target WASI, and will
enable us to remove existing uses of `rustup` at build-time.

`brew audit --new` fails because of the notability check, but I think we
can/should make an exception for this.

### CODE REVIEW

**Comment by cho-m** [comment]
I think there is a vendored copy in Rust and a wrapper around it https://github.com/rust-lang/rust/tree/1.82.0/src/tools/wasm-component-ld. It is currently built, but haven't checked if this was intended, e.g.
```console
❯ fd wasm-component-ld /opt/homebrew/opt/rust/
/opt/homebrew/opt/rust/lib/rustlib/aarch64-apple-darwin/bin/wasm-component-ld

❯ /opt/homebrew/opt/rust/lib/rustlib/aarch64-apple-darwin/bin/wasm-component-ld --version
wasm-component-ld 0.5.7
```

**Comment by carlocab** [comment]
Yep, I found that, but after testing it didn't seem to work. Not sure if I was just using it wrong.

**Comment by cho-m** [comment]
Having a separate formula is also fine.

I also wasn't sure if Rust was supposed to include the `wasm-component-ld`. I did see there was a PR to change build condition at https://github.com/rust-lang/rust/pull/130034

File: Formula/w/wasm-component-ld.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      volatile int x = 42;
+      int main(void) {
+        printf("the answer is %d", x);
+        return 0;
+      }
+    C
+    (testpath/"test.cc").write <<~CPP
+      #include <iostream>
+
+      // FIXME: we build without exceptions, but our C++ library still references them.
```
Comment (carlocab) [inline Formula/w/wasm-component-ld.rb:35]: Will be fixed in #196027.

File: Formula/w/wasi-runtimes.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ class WasiRuntimes < Formula
   depends_on "cmake" => :build
   depends_on "lld" => [:build, :test]
   depends_on "wasi-libc" => [:build, :test]
+  depends_on "wasm-component-ld" => [:build, :test]
```
Comment (carlocab) [inline Formula/w/wasi-runtimes.rb:25]: Looks like we need this so that toolchain features are correctly detected when configuring the libc++ build.

**Review by SMillerDev — APPROVED** [review]

**Pull request merged**